src/airstack/onchain_graph.py

```diff
@@ -0,0 +1,759 @@
+                    isPrimary
+                }
+                socials {
+                    dappName
+                    blockchain
+                    profileName
+                    profileImage
+                    profileTokenId
+                    profileTokenAddress
+                }
+                xmtp {
+                    isXMTPEnabled
+                }
+                }
+            }
+            }
+        }
+        }
+        """
+        poaps_data_response = None
+        recommended_users = existing_users.copy()
+        while True:
+            if poaps_data_response is None:
+                execute_query_client = self.api_client.create_execute_query_object(
+                    query=user_poaps_event_ids_query, variables={'user': address})
+                # Pagination #1: Fetch All POAPs
+                poaps_data_response = await execute_query_client.execute_paginated_query()
+
+            if poaps_data_response.error is None:
+                event_ids = [
+                    poap.get('eventId')
+                    for poap in poaps_data_response.data.get('Poaps', {}).get('Poap', [])
+                    if not poap.get('poapEvent', {}).get('isVirtualEvent')
+                ] if poaps_data_response.data and 'Poaps' in poaps_data_response.data and 'Poap' in poaps_data_response.data['Poaps'] else []
+                poap_holders_data_response = None
+                while True:
+                    if poap_holders_data_response is None:
+                        execute_query_client = self.api_client.create_execute_query_object(
+                            query=poaps_by_event_ids_query, variables={'eventIds': event_ids})
+                        # Pagination 2: Fetch all POAP Holders
+                        poap_holders_data_response = await execute_query_client.execute_paginated_query()
+
+                    if poap_holders_data_response.error is None:
+                        recommended_users = format_poaps_data(
+                            poap_holders_data_response.data.get(
+                                'Poaps', {}).get('Poap', []),
+                            recommended_users
+                        )
+
+                        if not poap_holders_data_response.has_next_page:
+                            break
+                        else:
+                            poap_holders_data_response = await poap_holders_data_response.get_next_page
+                    else:
+                        print("Error: ", poap_holders_data_response.error)
+                        break
+
+                if not poaps_data_response.has_next_page:
+                    break
+                else:
+                    poaps_data_response = await poaps_data_response.get_next_page
+            else:
+                print("Error: ", poaps_data_response.error)
+                break
+
+        return recommended_users
+
+    async def __fetch_farcaster_followings(self, address, existing_users=[]):
```

Comment:
There is a significant amount of code duplication in the methods that fetch data for different types (e.g., __fetch_farcaster_followings, __fetch_lens_followings, etc.). Consider refactoring these methods to reduce redundancy by creating a more generic method that accepts parameters for the specific type of data to fetch.